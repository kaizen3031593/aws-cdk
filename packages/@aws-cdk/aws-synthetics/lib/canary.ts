import { Alarm, AlarmProps, Metric, MetricOptions } from '@aws-cdk/aws-cloudwatch';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct, Duration, Resource, ResourceProps, IResource } from '@aws-cdk/core';
import { CfnCanary, Code } from '../lib';

/**
 * The rate specifies the rate that the Canary runs.
 *
 * The rate forms the expression string that is expected in the Canary definition
 * as part of the Schedule property. A valid expression has the syntax 'rate(number unit)'
 * where the unit can be minute, minutes, or hour. 'rate(0 minute) or rate(0 hour) is a
 * special value that causes the canary to run only once.
 *
 * If you need to use an expression that doesn't exist as a static member, you can
 * instantiate a 'Expression' object, e.g: 'new Expression('rate(15 minutes)')'.
 */
export class Rate {
  /**
   * The expression rate(1 minute)
   */
  public static readonly ONE_MINUTE = new Rate('rate(1 minute)');

  /**
   * The expression rate(5 minutes)
   */
  public static readonly FIVE_MINUTES = new Rate('rate(5 minutes)');

  /**
   * The expression rate(1 hour)
   */
  public static readonly ONE_HOUR = new Rate('rate(1 hour)');

  /**
   * The expression rate(0 minute)
   */
  public static readonly RUN_ONCE = new Rate('rate(0 minute)');

  /**
   * The expression that is expected by the Canary resource
   */
  public readonly expression: string;

  constructor(expression: string) {
    this.verifyExpression(expression);
    this.expression = expression;
  }

  /**
   * The rate expressed as a string
   */
  public toString(): string {
    return this.expression;
  }

  private verifyExpression(expression: string) {
    const exp = expression.split(' ');
    const rateAndNumber = exp[0].split('(');
    if (exp.length !== 2 ||
      rateAndNumber.length !== 2 ||
      rateAndNumber[0] !== 'rate' ||
      isNaN(Number(rateAndNumber[1])) ||
      (exp[1] !== 'minute)' && exp[1] !== 'minutes)' && exp[1] !== 'hour)')) {
      throw new Error('Expression must follow the syntax \'rate(number unit)\'');
    }
    if (exp[1] === 'hour)' && Number(rateAndNumber[1]) > 1 ||
        exp[1] === 'minutes)' && Number(rateAndNumber[1]) > 60)  {
      throw new Error('Expression must not be greater than 1 hour');
    }
  }
}

/**
 * Options for a canary
 */
export interface CanaryOptions extends ResourceProps {
  /**
   * The location to store the data of the canary runs.
   *
   * @default - A new s3 bucket will be created.
   */
  readonly artifactS3Location?: string;
  /**
   * Canary execution role.
   *
   * This is the role that will be assumed by the canary upon execution.
   * It controls the permissions that the canary will have. The Role must
   * be assumable by the 'lambda.amazonaws.com' service principal.
   *
   * The default Role automatically has permissions granted for Canary execution.
   * If you provide a Role, you must add the relevant policies yourself.
   *
   * The relevant policies are "s3:PutObject", "s3:GetBucketLocation", "s3:ListAllMyBuckets",
   * "cloudwatch:PutMetricData", "logs:CreateLogGroup", "logs:CreateLogStream", "logs:PutLogEvents".
   *
   * @default - A unique role will be generated for this canary.
   * Both supplied and generated roles can always be changed by calling 'addToRolePolicy'.
   */
  readonly role?: iam.IRole;

  /**
   * How many seconds the canary should run before timing out.
   *
   * @default - same amount as rate
   */
  readonly timeout?: Duration;

  /**
   * How many seconds the canary will make regular runs according to the schedule in 'expression'.
   *
   * The default of 0 seconds means that the canary will continue to make runs until you stop it.
   *
   * @default Duration.seconds(0)
   */
  readonly lifetime?: Duration;

  /**
   * How often the canary will run during the duration. The syntax for expression is 'rate(number unit)'
   * where unit can be 'minute', 'minutes', or 'hour'. You can specify a frequency between 'rate(1 minute)'
   * and 'rate(1 hour)'.
   *
   * The default of 'rate(0 minute)' specifies that the canary will run only once when it is started.
   *
   * @default 'rate(0 minute)'
   */
  readonly rate?: Rate;

  /**
   * Whether or not the canary should start after creation.
   *
   * @default true
   */
  readonly enableCanary?: boolean;

  /**
   * How many days should successful runs be retained
   *
   * @default Duration.days(31)
   */
  readonly successRetentionPeriod?: Duration;

  /**
   * How many days should failed runs be retained
   *
   * @default Duration.days(31)
   */
  readonly failureRetentionPeriod?: Duration;

  /**
   * If the endpoint that the canary is testing is inside of a vpc, then you must specify
   * which vpc.
   *
   * @default none
   */
  readonly vpc?: ec2.IVpc;
}

/**
 * Properties of a canary
 */
export interface CanaryProps extends CanaryOptions {
  /**
   * The handler of the code.
   */
  readonly handler: string;

  /**
   * The canary code. Currently only supported by inline code. In the future,
   * it will be possible to specify an asset or an s3 bucket where the code is.
   */
  readonly code: Code;

  /**
   * The name of the canary.
   *
   * @default - A unique physical ID will be generated for you and used as the canary's name.
   */
  readonly canaryName: string;
}

export interface ICanary extends IResource {
  // Should extend iam.IGrantable, ec2.IConnectable
  // Should have all the properties and method signatures associated with a canary
}

/**
 * Represents a Canary defined outside of this stack.
 */
export interface CanaryAttributes {
  // All you need to find a canary from outside the stack
}

/**
 * Base of a canary
 */
export abstract class CanaryBase extends Resource implements ICanary {
}

/**
 * The canary.
 */
export class Canary extends CanaryBase {

  /**
   * Return the given named metric for all canaries
   */
  public static metricAll(metricName: string, props?: MetricOptions): Metric {
    return new Metric({
      namespace: 'CloudWatchSynthetics',
      metricName,
      ...props,
    });
  }

  /**
   * Metric for SuccessPercent across all canaries
   *
   * @default average over 5 minutes
   */
  public static metricAllSuccessPercent(props?: MetricOptions): Metric {
    return this.metricAll('SuccessPercent', { statistic: 'avg', ...props });
  }

  /**
   * Metric for Duration across all canaries
   *
   * @default average over 5 minutes
   */
  public static metricAllDuration(props?: MetricOptions): Metric {
    return this.metricAll('Duration', { statistic: 'avg', ...props });
  }

  /**
   * Execution role associated with this Canary.
   */
  public readonly role: iam.IRole;

  /**
   * VPC network to place Canary network interfaces
   *
   * Specify this if the Canary needs to access resources in a VPC.
   *
   * @default - Canary does not need to access a VPC.
   */
  public readonly vpc?: ec2.IVpc;

  /**
   * Canary ID
   * @attribute
   */
  public readonly canaryId: string;

  /**
   * Canary State
   * @attribute
   */
  public readonly canaryState: string;

  /**
   * Canary Name
   * @attribute
   */
  public readonly canaryName: string;

  constructor(scope: Construct, id: string, props: CanaryProps) {
    super(scope, id);

    const s3Location = props.artifactS3Location || new s3.Bucket(this, 'ServiceBucket').s3UrlForObject();

    // Created role will need these policies to run the Canary.
    const policy = new iam.PolicyDocument({
      statements: [new iam.PolicyStatement({
        resources: ['*'],
        actions: [
          's3:PutObject',
          's3:GetBucketLocation',
          's3:ListAllMyBuckets',
          'cloudwatch:PutMetricData',
          'logs:CreateLogGroup',
          'logs:CreateLogStream',
          'logs:PutLogEvents',
        ],
      })],
    });
    const inlinePolicies = { canaryPolicy : policy };

    this.role = props.role || new iam.Role(this, 'ServiceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      inlinePolicies,
    });

    const duration = props.lifetime || Duration.seconds(0);
    const expression = props.rate || Rate.ONE_MINUTE;
    const timeout = props.timeout || Duration.seconds(60);

    const code = props.code.bind(this);

    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
      artifactS3Location: s3Location,
      executionRoleArn: this.role.roleArn,
      startCanaryAfterCreation: props.enableCanary ?? true,
      runtimeVersion: 'syn-1.0',
      name: this.verifyName(props.canaryName),
      runConfig: { timeoutInSeconds: timeout.toSeconds()},
      schedule: { durationInSeconds: String(duration.toSeconds()), expression: expression.toString() },
      failureRetentionPeriod: props.failureRetentionPeriod?.toDays(),
      successRetentionPeriod: props.successRetentionPeriod?.toDays(),
      code: {
        handler: this.verifyHandler(props.handler),
        script: code.inlineCode,
        s3Bucket: code.s3Location?.bucketName,
        s3Key: code.s3Location?.objectKey,
        s3ObjectVersion: code.s3Location?.objectVersion,
      },
    });
    resource.node.addDependency(this.role);

    // this.canaryState = resource.getAtt('state').toString();
    this.canaryId = resource.attrId;
    this.canaryState = resource.attrState;
    this.canaryName = this.getResourceNameAttribute(resource.ref);
  }

  /**
   * Returns a new metric for the canary
   *
   * @default avg over 5 minutes
   */
  public metric(metricName: string, props?: MetricOptions): Metric {
    return new Metric({
      metricName,
      namespace: 'CloudWatchSynthetics',
      dimensions: { CanaryName: this.canaryName },
      statistic: 'avg',
      ...props,
    }).attachTo(this);
  }

  /**
   * Returns a Duration metric for the canary
   *
   * @default avg over 10 minutes
   */
  public metricDuration(props?: MetricOptions): Metric {
    return this.metric('Duration', props);
  }

  /**
   * Returns a Success Percent metric for the canary
   *
   * @default avg over 10 minutes
   */
  public metricSuccessPercent(props?: MetricOptions): Metric {
    return this.metric('SuccessPercent', props);
  }

  /**
   * Returns Failed metric for the canary
   *
   * @default avg over 10 minutes
   */
  public metricFailed(props?: MetricOptions): Metric {
    return this.metric('Failed', props);
  }

  /**
   * Returns an alarm for the canary
   */
  public createAlarm(id: string, props: AlarmProps): Alarm {
    return new Alarm(this, id, props);
  }

  /**
   * Verifies if the given handler ends in '.handler'. Returns the handler if successful and
   * throws an error if not.
   *
   * @param handler - the handler given by the user
   */
  private verifyHandler(handler: string): string {
    if (handler.split('.').length !== 2 || handler.split('.')[1] !== 'handler') {
      throw new Error('Canary Handler must end in \'.handler\'');
    }
    if (handler.length > 21) {
      throw new Error('Canary Handler must be less than 21 characters.');
    }
    return handler;
  }

  /**
   * Verifies that the name fits the regex expression: ^[0-9a-z_\-]+$
   * Returns the name if successful and throws an error if not.
   *
   * @param name - the given name of the canary
   */
  private verifyName(name: string): string {
    const regex = new RegExp('^[0-9a-z_\-]+$');
    if (!regex.test(name)) {
      throw new Error('Canary Name must be lowercase, numbers, hyphens, or underscores (no spaces).');
    }
    if (name.length > 21) {
      throw new Error('Canary Name must be less than 21 characters');
    }
    return name;
  }
}
