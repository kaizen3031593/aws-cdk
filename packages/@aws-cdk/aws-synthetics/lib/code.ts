import * as s3 from '@aws-cdk/aws-s3';
import * as s3_assets from '@aws-cdk/aws-s3-assets';
import * as cdk from '@aws-cdk/core';
import * as path from 'path';
import * as fs from 'fs';
// import * as archiver from 'archiver';

/**
 * The code of the canary. Currently only supports inline code.
 */
export abstract class Code {
  /**
   * Specify Code parameter from inline.
   *
   * @returns `CanaryInlineCode` with inline code.
   * @param code The actual handler code (limited to 4KiB)
   */
  public static fromInline(code: string): InlineCode {
    return new InlineCode(code);
  }

  public static fromAsset(path: string, options?: s3_assets.AssetOptions): AssetCode {
    return new AssetCode(path, options);
  }

  public static fromBucket(bucket: s3.IBucket, key: string, objectVersion?: string): S3Code {
    return new S3Code(bucket, key, objectVersion);
  }

  /**
   * Called when the canary is initialized to allow this object to bind
   * to the stack, add resources and have fun.
   *
   * @param scope The binding scope. Don't be smart about trying to down-cast or
   * assume it's initialized. You may just use it as a construct scope.
   */
  public abstract bind(scope: cdk.Construct): CodeConfig;

  public bindToResource(_resource: cdk.CfnResource, _options?: ResourceBindOptions) {
    return;
  }

  public abstract readonly isInline: boolean;
}

/**
 * Canary code from a local directory.
 */
export class AssetCode extends Code {
  public readonly isInline = false;
  private asset?: s3_assets.Asset;

  /**
   * @param path The path to the asset file or directory.
   */
  constructor(public readonly path: string, private readonly options: s3_assets.AssetOptions = { }) {
    super();
  }

  public bind(scope: cdk.Construct): CodeConfig {
    // If the same AssetCode is used multiple times, retain only the first instantiation.
    if (!this.asset) {
      if (path.extname(this.path) === ''){
        this.verifyPath();
      }
      this.asset = new s3_assets.Asset(scope, 'Code', {
        path: path.join(this.path),
        ...this.options,
      });
    }

    if (!this.asset.isZipArchive) {
      throw new Error(`Asset must be a .zip file or a directory (${this.path})`);
    }

    return {
      s3Location: {
        bucketName: this.asset.s3BucketName,
        objectKey: this.asset.s3ObjectKey,
      },
    };
  }

  private verifyPath() {
    const getDirectories = fs.readdirSync(this.path, { withFileTypes: true}).filter(dirent => dirent.isDirectory()).map(dirent => dirent.name);
    if (!getDirectories.includes('nodejs')) {
      throw new Error('Local directory must include the folder structure nodejs/node_modules');
    }
    const newPath = path.join(this.path, 'nodejs');
    const getNewDirectories = fs.readdirSync(newPath, { withFileTypes: true}).filter(dirent => dirent.isDirectory()).map(dirent => dirent.name);
    if (!getNewDirectories.includes('node_modules')) {
      throw new Error('Local directory must include the folder structure nodejs/node_modules');
    }
  }

  // private createStructure() {
  //   const internalZipPath = 'nodejs/node_modules';
  //   // makes the directory structure if it does not exist
  //   fs.mkdirSync(path.join(this.path, internalZipPath), { recursive: true });
  //   const filesToMove = fs.readdirSync(this.path);
  //   filesToMove.forEach((file) => {
  //     const pathToFile = path.join(this.path, file);
  //     if (path.extname(file) == '.js'){
  //       fs.copyFileSync(pathToFile, path.join(this.path,internalZipPath,file));
  //       // fs.rename(pathToFile, path.join(this.path, internalZipPath, file), (err) => {
  //       //   if (err) throw err;
  //       // });
  //     }
  //   })
  // }

  // private createZip() {
  //   const internalZipPath = 'nodejs/node_modules';

  //   var output = fs.createWriteStream(this.path + '/nodejs.zip');
  //   var archive = archiver('zip', {});
  //   archive.on('warning', function(err: any) {
  //     if (err.code === 'ENOENT') {
  //       // log warning
  //     } else {
  //       // throw error
  //       throw err;
  //     }
  //   });

  //   archive.on('error', function(err: Error) {
  //     throw err;
  //   });

  //   archive.on('finish', function() {
  //     console.log('canaryScript is zipped');
  //   });
  //   archive.pipe(output);
    
  //   const filesToAdd = fs.readdirSync(this.path);

  //   filesToAdd.forEach((file) => {
  //     const pathToFile = path.join(this.path, file);
  //     archive.file(pathToFile, {
  //       name: path.join(internalZipPath, file),
  //     }); 
  //   });
  //   archive.finalize();
  // }

  public bindToResource(resource: cdk.CfnResource, options: ResourceBindOptions = { }) {
    if (!this.asset) {
      throw new Error('bindToResource() must be called after bind()');
    }

    const resourceProperty = options.resourceProperty || 'Code';

    // https://github.com/aws/aws-cdk/issues/1432
    this.asset.addResourceMetadata(resource, resourceProperty);
  }
}

export interface ResourceBindOptions {
  readonly resourceProperty?: string;
}

/**
 * Codeconfig
 */
export interface CodeConfig {
  /**
   * The location of the code in S3 (mutually exclusive with `inlineCode`).
   *
   * @default none
   */
  readonly s3Location?: s3.Location;

  /**
   * Inline code (mutually exclusive with `s3Location`).
   *
   * @default none
   */
  readonly inlineCode?: string;
}

export class S3Code extends Code {
  public readonly isInline = false;
  private bucketName: string;

  constructor(bucket: s3.IBucket, private key: string, private objectVersion?: string) {
    super();

    if (!bucket.bucketName) {
      throw new Error('bucketName is undefined for the provided bucket');
    }

    this.bucketName = bucket.bucketName;
  }

  public bind(_scope: cdk.Construct): CodeConfig {
    return {
      s3Location: {
        bucketName: this.bucketName,
        objectKey: this.key,
        objectVersion: this.objectVersion,
      },
    };
  }
}

/**
 * InlineCode
 */
export class InlineCode extends Code {
  public readonly isInline = true;

  constructor(private code: string) {
    super();

    if (code.length === 0) {
      throw new Error('Canary inline code cannot be empty');
    }

    if (code.length > 4096) {
      throw new Error('Canary source is too large, must be <= 4096 but is ' + code.length);
    }
  }

  /**
   * Bind to the resource
   *
   * @param _scope - the construct
   */
  public bind(_scope: cdk.Construct): CodeConfig {
    return {
      inlineCode: this.code,
    };
  }
}