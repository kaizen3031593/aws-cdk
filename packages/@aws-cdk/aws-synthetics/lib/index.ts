export * from './canary';
export * from './code';

// AWS::Synthetics CloudFormation Resources:
export * from './synthetics.generated';
