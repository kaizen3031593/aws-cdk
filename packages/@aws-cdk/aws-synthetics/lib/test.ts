import * as cdk from '@aws-cdk/core';

export abstract class Test {
  
}