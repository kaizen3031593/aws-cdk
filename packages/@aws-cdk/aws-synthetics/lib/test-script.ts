import * as s3 from '@aws-cdk/aws-s3';
import { Construct, hashValidator } from '@aws-cdk/core';
import { Code } from './code';

/**
 * Specify a test that the canary should run
 */
export class Test {
  /**
   * Specify a custom test with your own code
   *
   * @returns `Test` associated with the specified Code object
   * @param code The script you want the canary to run
   * @param handler The handler of the code
   */
  public static custom(scope: Construct, options: CustomOptions): Test {
    const codeConfig = options.code.bind(scope);
    return new Test({
      handler: options.handler,
      inlineCode: codeConfig.inlineCode,
      s3Location: codeConfig.s3Location,
    });
  }

  /**
   * Use this method to access the template for heartBeat monitoring. This test will hit the specified url and report back
   * information on the health of the website.
   *
   * @returns `Test` with the heartBeat template as the Code
   * @param url The url of the website you want to test
   *
   * TODO: implement
   */
  public static heartBeat(url: string){
    return new Test({
      handler: 'index.handler',
      inlineCode: new HeartBeatTemplate(url).inlineCode,
    });
  }

  /**
   * Use this method to access the template for api Endpoint monitoring. This test will hit the endpoint specified and report
   * back what it receives.
   *
   * @param url The url of the api endpoint you want to test
   * @param options Options to customize the api endpoint template
   *
   * TODO: implement
   */
  //public static apiEndpoint(url: string, options: apiOptions){}

  /**
   * Use this method to access the template for a broken link checker. This test will check a specified number of links and report
   * back the first broken link found.
   *
   * @param url The starting url to check
   * @param options Options to customzie the template
   *
   * TODO: implement
   */
  //public static brokenLink(url: string, options: linkOptions){}

  private constructor(public readonly config: TestConfig){}
}

/**
 * Configuration property of the test class
 */
export interface TestConfig {
  /**
   * The location of the code in S3 (mutually exclusive with `inlineCode`).
   */
  readonly s3Location?: s3.Location;

  /**
   * Inline code (mutually exclusive with `s3Location`).
   */
  readonly inlineCode?: string;

  /**
   * The handler of the code
   */
  readonly handler: string;
}

/**
 * Options for specifying a custom test
 */
export interface CustomOptions {
  /**
   * The code of the canary script
   */
  readonly code: Code,

  /**
   * The handler for the code
   */
  readonly handler: string,
}

export class HeartBeatTemplate {
  readonly inlineCode: string;

  constructor(url: string) {
    this.inlineCode = `var synthetics = require('Synthetics');
      const log = require('SyntheticsLogger');
      
      const pageLoadBlueprint = async function () {
      
          const URL = "${url}";
      
          let page = await synthetics.getPage();
          const response = await page.goto(URL, {waitUntil: 'domcontentloaded', timeout: 30000});
          //Wait for page to render.
          //Increase or decrease wait time based on endpoint being monitored.
          await page.waitFor(15000);
          await synthetics.takeScreenshot('loaded', 'loaded');
          let pageTitle = await page.title();
          log.info('Page title: ' + pageTitle);
          if (response.status() !== 200) {
              throw "Failed to load page!";
          }
      };
      
      exports.handler = async () => {
          return await pageLoadBlueprint();
      };`;
  }
}