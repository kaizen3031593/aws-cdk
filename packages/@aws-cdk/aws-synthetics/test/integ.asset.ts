import * as cdk from '@aws-cdk/core';
import * as synth from '../lib';
import * as path from 'path';

const name = 'asset-canary-13';
/*
 * Stack verification steps:
 *
 * -- aws synthetics get-canary --name asset-test has a state of 'RUNNING'
 */
const app = new cdk.App();
const stack = new cdk.Stack(app, name);

new synth.Canary(stack, name, {
  canaryName: name,
  code: synth.Code.fromAsset(path.join(__dirname, 'canary')),
  handler: 'canary.handler',
});

app.synth();
