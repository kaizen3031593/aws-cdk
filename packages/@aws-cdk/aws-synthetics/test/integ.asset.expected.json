{
  "Resources": {
    "assetcanary13ServiceBucket1362D80C": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "assetcanary13ServiceRole5D462A3D": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:GetBucketLocation",
                    "s3:ListAllMyBuckets",
                    "cloudwatch:PutMetricData",
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents"
                  ],
                  "Effect": "Allow",
                  "Resource": "*"
                }
              ],
              "Version": "2012-10-17"
            },
            "PolicyName": "canaryPolicy"
          }
        ]
      }
    },
    "assetcanary13665D3B8D": {
      "Type": "AWS::Synthetics::Canary",
      "Properties": {
        "ArtifactS3Location": {
          "Fn::Join": [
            "",
            [
              "s3://",
              {
                "Ref": "assetcanary13ServiceBucket1362D80C"
              }
            ]
          ]
        },
        "Code": {
          "Handler": "canary.handler",
          "S3Bucket": {
            "Ref": "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602S3Bucket513E3DB1"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602S3VersionKey398BC680"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602S3VersionKey398BC680"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "assetcanary13ServiceRole5D462A3D",
            "Arn"
          ]
        },
        "Name": "asset-canary-13",
        "RunConfig": {
          "TimeoutInSeconds": 60
        },
        "RuntimeVersion": "syn-1.0",
        "Schedule": {
          "DurationInSeconds": "0",
          "Expression": "rate(1 minute)"
        },
        "StartCanaryAfterCreation": true
      },
      "DependsOn": [
        "assetcanary13ServiceRole5D462A3D"
      ]
    }
  },
  "Parameters": {
    "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602S3Bucket513E3DB1": {
      "Type": "String",
      "Description": "S3 bucket for asset \"df72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602\""
    },
    "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602S3VersionKey398BC680": {
      "Type": "String",
      "Description": "S3 key for asset version \"df72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602\""
    },
    "AssetParametersdf72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602ArtifactHash9B840AD0": {
      "Type": "String",
      "Description": "Artifact hash for asset \"df72d06c2d37a55cb680d0262258d53f82692ebb572d35ce0ffae76a51f3a602\""
    }
  }
}